cadc-cdp/src/main/java/ca/nrc/cadc/cred/CertUtil.java

```diff
@@ -129,140 +144,119 @@ public class CertUtil
         Date beforeDate = date.getTime();
-        for (X509Certificate currentCert : chain.getChain())
-        {
-            if (beforeDate.before(currentCert.getNotBefore()))
-            {
+        for (X509Certificate currentCert : chain.getChain()) {
+            if (beforeDate.before(currentCert.getNotBefore())) {
                 beforeDate = currentCert.getNotBefore();
             }
         }
-        certGen.setNotBefore(beforeDate);
 
-        // End date.
+        // end date
         // If hours = 0, then cert lifetime is set to that of user cert
-        if (lifetime <= 0)
-        {
+        date = new GregorianCalendar(TimeZone.getTimeZone("GMT"));
+        Date afterDate = date.getTime();
+        if (lifetime <= 0) {
             // set the validity of certificates as the minimum
             // of the certificates in the chain
-            Date afterDate = issuerCert.getNotAfter();
-            for (X509Certificate currentCert : chain.getChain())
-            {
-                if (afterDate.after(currentCert.getNotAfter()))
-                {
+            afterDate = issuerCert.getNotAfter();
+            for (X509Certificate currentCert : chain.getChain()) {
+                if (afterDate.after(currentCert.getNotAfter())) {
                     afterDate = currentCert.getNotAfter();
                 }
             }
-            certGen.setNotAfter(afterDate);
-        }
-        else
-        {
+        } else {
             // check the validity of the signing certificate
             date.add(Calendar.MINUTE, 5);
             date.add(Calendar.SECOND, lifetime);
-            for (X509Certificate currentCert : chain.getChain())
-            {
+            for (X509Certificate currentCert : chain.getChain()) {
                 currentCert.checkValidity(date.getTime());
             }
-
-            certGen.setNotAfter(date.getTime());
+            afterDate = date.getTime();
         }
 
-        certGen.setPublicKey(csr.getPublicKey());
-        // TODO: should be able to get signature algorithm from the csr, but... obtuse
-        certGen.setSignatureAlgorithm(DEFAULT_SIGNATURE_ALGORITHM);
-
-        // extensions
-        // add ProxyCertInfo extension to the new cert
-
-        certGen.addExtension(X509Extensions.KeyUsage, true, 
-            new KeyUsage(KeyUsage.digitalSignature | KeyUsage.keyEncipherment));
-
-        certGen.addExtension(X509Extensions.AuthorityKeyIdentifier,
-            false, new AuthorityKeyIdentifierStructure(issuerCert));
+        X509v3CertificateBuilder certGen = new X509v3CertificateBuilder(issuer,
+                serial, beforeDate, afterDate, subject, csr.getSubjectPublicKeyInfo());
 
-        certGen.addExtension(X509Extensions.SubjectKeyIdentifier,
-            false, new SubjectKeyIdentifierStructure(csr.getPublicKey("BC")));
+        try {
+            // add ProxyCertInfo extension
+            ASN1EncodableVector policy = new ASN1EncodableVector();
```

Comment:
It would be nice to write up a summary of the steps bellow in the unlikely case that we need to maintain this in the future.